components/server/src/raft_engine_switch.rs

```diff
@@ -137,22 +50,79 @@ pub fn check_and_dump_raft_db(
     }
     drop(tx);
     info!("Scanned all region id and waiting for dump");
-    for t in threads {
+    for t in workers {
         t.join().unwrap();
     }
-    engine.sync().unwrap();
+    target.sync().unwrap();
     info!(
         "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
         count_region,
         count_size.load(Ordering::Relaxed),
         consumed_time.saturating_elapsed(),
     );
+}
 
-    rename_to_tmp_dir(&raftdb_path, &dirty_raftdb_path);
-    remove_tmp_dir(&dirty_raftdb_path);
+pub fn dump_raft_engine_to_raftdb(source: &RaftLogEngine, target: &RocksEngine, threads: usize) {
+    check_raft_db_is_empty(target);
+
+    let count_size = Arc::new(AtomicUsize::new(0));
+    let mut count_region = 0;
+    let mut workers = vec![];
+    let (tx, rx) = unbounded();
+    for _ in 0..threads {
+        let source = source.clone();
+        let target = target.clone();
+        let count_size = count_size.clone();
+        let rx = rx.clone();
+        let t = std::thread::spawn(move || {
+            run_dump_raft_engine_worker(&rx, &source, &target, &count_size);
+        });
+        workers.push(t);
+    }
+
+    info!("Start to scan raft log from RaftLogEngine and dump into RocksEngine");
+    let consumed_time = tikv_util::time::Instant::now();
+    // Seek all region id from RaftLogEngine and send them to workers.
+    for id in source.raft_groups() {
+        tx.send(id).unwrap();
+        count_region += 1;
+    }
+    drop(tx);
+
+    info!("Scanned all region id and waiting for dump");
+    for t in workers {
+        t.join().unwrap();
+    }
+    target.sync().unwrap();
+    info!(
+        "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
+        count_region,
+        count_size.load(Ordering::Relaxed),
+        consumed_time.saturating_elapsed(),
+    );
+}
+
+fn check_raft_engine_is_empty(engine: &RaftLogEngine) {
+    assert!(
+        engine.raft_groups().is_empty(),
+        "Cannot transfer data from RaftDb to non-empty Raft Engine."
+    );
+}
+
+fn check_raft_db_is_empty(engine: &RocksEngine) {
+    let mut count = 0;
+    engine
+        .scan(b"", &[0xFF, 0xFF], false, |_, _| {
+            count += 1;
+            Ok(true)
```

Comment:
```suggestion
            Ok(false)
```
return directly if there is any.